### PR TITLE
Check for AsRef/AsMut arguments in wrong_self_convention

### LINKED ISSUE #451: False wrong_self_convention positive with AsRef bound
Example from Rust-Bio:

```
src/io/fastq.rs:178:36: 178:47 warning: methods called `to_*` usually take self by reference; consider choosing a less ambiguous name, #[warn(wrong_self_convention)] on by default
src/io/fastq.rs:178     pub fn to_file<P: AsRef<Path>>(path: P) -> io::Result<Self> {
                                                       ^~~~~~~~~~~
src/io/fastq.rs:178:36: 178:47 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#wrong_self_convention
```

This is not &self anyway, so `wrong_self_convention` seems a bit out of place. Apart from that, `AsRef<T>` should be regarded equal to `&T`

The same would happen with `AsMut` vs, `&mut`, I presume.

### PR DESCRIPTION
**Pull request by llogiq**
This fixes #451

### CODE REVIEW

File: clippy_lints/src/methods.rs [hunk 1]
```diff
@@ -1398,6 +1403,45 @@ impl SelfKind {
...
+            })
+        })
+    })
+}
+
+fn is_ty(ty: &hir::Ty, self_ty: &hir::Ty) -> bool {
+    match (&ty.node, &self_ty.node) {
+        (&hir::TyPath(hir::QPath::Resolved(_, ref ty_path)), &hir::TyPath(hir::QPath::Resolved(_, ref self_ty_path))) => {
+            ty_path.segments.iter().rev().map(|seg| seg.name).zip(
+                self_ty_path.segments.iter().rev().map(|seg| seg.name)).all(|(l, r)| l == r)
```
Comment (oli-obk) [inline clippy_lints/src/methods.rs]: This should use Iterator's eq method. And we need an issue open about linting this pattern ;)

Comment (Manishearth) [inline clippy_lints/src/methods.rs]: `.eq()`?

File: clippy_lints/src/methods.rs [hunk 2]
```diff
@@ -1398,6 +1403,45 @@ impl SelfKind {
     }
 }
 
+fn is_astrait(ty: &hir::Ty, self_ty: &hir::Ty, generics: &hir::Generics, name: &[&str]) -> bool {
```
Comment (Manishearth) [inline clippy_lints/src/methods.rs]: doc comment please, the name is kinda confusing

File: clippy_lints/src/methods.rs [hunk 3]
```diff
@@ -10,7 +10,7 @@ use syntax::codemap::Span;
 use utils::{get_trait_def_id, implements_trait, in_external_macro, in_macro, is_copy, match_path, match_trait_method,
             match_type, method_chain_args, return_ty, same_tys, snippet, span_lint, span_lint_and_then,
             span_note_and_lint, walk_ptrs_ty, walk_ptrs_ty_depth, last_path_segment, single_segment_path,
-            match_def_path, is_self, is_self_ty, iter_input_pats};
+            match_def_path, is_self, is_self_ty, iter_input_pats, match_path_old};
```
Comment (mcarton) [inline clippy_lints/src/methods.rs]: Why do we have a function named `*_old`? Oo

File: clippy_lints/src/methods.rs (hunk 2 above)
Comment (mcarton) [inline clippy_lints/src/methods.rs]: Why is this not `is_trait` and in `utils` btw?

File: clippy_lints/src/methods.rs (hunk 1 above)
Comment (llogiq) [inline clippy_lints/src/methods.rs]: Ok, will change.

File: clippy_lints/src/methods.rs (hunk 3 above)
Comment (llogiq) [inline clippy_lints/src/methods.rs]: Because there is a new one doing a slightly different thing...

File: clippy_lints/src/methods.rs (hunk 2 above)
Comment (llogiq) [inline clippy_lints/src/methods.rs]: I'll rename to `is_as_ref_or_mod_trait`. And it's kinda specific to this lint, what with the self test.

File: clippy_lints/src/methods.rs (hunk 1 above)
Comment (llogiq) [inline clippy_lints/src/methods.rs]: TIL.

**Pull request merged**